testbed/core/views.py

```diff
@@ -72,3 +82,200 @@ def index(request):
+    if redirect_uri not in application.redirect_uris:
+        if application.redirect_uris:
+            application.redirect_uris = f"{application.redirect_uris} {redirect_uri}"
+        else:
+            application.redirect_uris = redirect_uri
+        application.save()
+        logger.info(f"Added {redirect_uri} to redirect URIs for application {application.client_id}")
+    
+    # Generate a secure random state parameter and store it in the session
+    state = generate_secure_state()
+    store_state_in_session(request, state)
+    
+    # Build the authorization URL
+    params = {
+        'client_id': application.client_id,
+        'response_type': 'code',
+        'scope': 'activitypub_account_portability',
+        'redirect_uri': redirect_uri,
+        'state': state
+    }
+    
+    # Convert params to URL query string
+    query_string = '&'.join([f"{key}={value}" for key, value in params.items()])
+    auth_url = f"{reverse('oauth2_provider:authorize')}?{query_string}"
+    
+    return redirect(auth_url)
+
+@login_required
+def test_error_view(request):
+    """
+    Test view to simulate an OAuth error.
+    This renders the error template with a sample error.
+    """
+    error = {
+        'error': 'invalid_request',
+        'description': 'This is a test error to preview the error template.',
+        'uri': None
+    }
+    
+    return render(request, 'oauth2_provider/error.html', {'error': error})
+
+
+# Token Exchange View
+@login_required
+def test_token_exchange_view(request):
+    """
+    Demonstrate exchanging an authorization code for an access token.
+    
+    This view simulates what a destination service would do after receiving 
+    an authorization code. It makes a token request to the token endpoint
+    using the authorization code, client credentials, and redirect URI.
+    """
+    # Get query parameters from the callback
+    code = request.GET.get('code')
+    state = request.GET.get('state')
+    error = request.GET.get('error')
+    
+    context = {
+        'code': code,
+        'state': state,
+        'error': error,
+        'token_response': None,
+        'token_error': None,
+    }
+    
+    # If we have an error or no code, don't attempt token exchange
+    if error or not code:
+        context['token_error'] = "Cannot exchange token: no valid authorization code provided"
+        return render(request, 'oauth_token_exchange.html', context)
+    
+    # Get the application for client credentials
+    application = get_user_application(request.user)
+    
+    # Prepare token request parameters
+    token_url = f"{request.scheme}://{request.get_host()}/oauth/token/"
+    redirect_uri = f"{request.scheme}://{request.get_host()}/callback"
+    
+    # Prepare the data for token exchange
+    # This follows the OAuth 2.0 specification for token requests
+    token_data = {
+        'grant_type': 'authorization_code',
+        'code': code,
+        'redirect_uri': redirect_uri,
```

Comment:
question whether the redirect_uri is supposed to be a parameter on the token exchange request...